csrc/id_model/id_model.cpp

```diff
@@ -925,17 +946,87 @@ Expr* findMatchingExpr(
       continue;
     }
 
+    // For the final loop promotion map, we want to find
+    // promotions within the same loop groups. Note that that's
+    // guaranteed when a new domain is replayed instead of reusing an
+    // existing domain.
+    if (require_loop_mapped_promotion) {
+      if (!loop_graph.disjointExprSets().permissiveAreMapped(
+              iel_expr->front(), maybe_promoted_input_use_group->front())) {
+        continue;
+      }
+      // This is just an extra sanity check. Make sure all exprs in
+      // the use group are mapped
+      NVF_ERROR(
+          std::all_of(
+              maybe_promoted_input_use_group->vector().begin(),
+              maybe_promoted_input_use_group->vector().end(),
+              [&](Expr* iel_use) {
+                return loop_graph.disjointExprSets().permissiveAreMapped(
+                    iel_expr->front(), iel_use);
+              }),
+          "Not all mapped: ",
+          nvfuser::toString(iel_expr),
+          "\n",
+          nvfuser::toString(maybe_promoted_input_use_group));
+    }
+
     return maybe_promoted_input_use;
   }
 
   return nullptr;
 }
 
+// When propagating loop promotions from inputs to outputs of an IEL
+// expr, we can't blindly apply loop promotion when all of the input
+// domains are loop mapped with the outputs.
+//
+// i.e. if we have the inlined domains from:
+// T2[i0*i1] pa(1) = T0[i0*b1]ca(1) + T1[i0*i1]ca(1)
+// The inlined loop group would be:
+//
+// i0, i1, b1, i0*i1, b0*i1
```

Comment:
Thanks, you're right. The example is not correct. Let me update.

Reply:
I do not have a concrete counter-example, but to me this sounds like "I found this unexpected behavior when there is this pattern. So I just do a pattern matching to special handle it." This sounds like just hiding the problem deeper, instead of solving the problem from its root. I am gaining more confidence that we should remove forwarding and find a better way to handle broadcast IDs created in the fusion.

To be clear, I am not asking any change for now. I am more talking about what we should do in the long term.

Reply:
Updated the comment. Please let me know if it makes sense.

I completely agree that this is one of the ugly complications of the Loop/Permissive graphs.

Reply:
It makes sense and very clear! Thank you for writing the explanation.